sound/pci/hda/patch_hdmi.c

```diff
@@ -1642,30 +1650,86 @@ static void hdmi_present_sense_via_verbs(struct hdmi_spec_per_pin *per_pin,
 	snd_hda_power_down_pm(codec);
 }
 
+#define INTEL_SILENTSTR_RATE		48000
+#define INTEL_SILENTSTR_CHANNELS		2
+#define INTEL_SILENTSTR_FORMAT		SNDRV_PCM_FORMAT_S16_LE
+#define INTEL_SILENTSTR_TAG		0xf
+
 static void silent_stream_enable(struct hda_codec *codec,
-				struct hdmi_spec_per_pin *per_pin)
+				 struct hdmi_spec_per_pin *per_pin)
 {
-	unsigned int newval, oldval;
+	struct hdmi_spec_per_cvt *per_cvt = NULL;
+	struct hdmi_spec *spec = codec->spec;
+	int cvt_idx, pin_idx, err;
+	unsigned int format;
 
-	codec_dbg(codec, "hdmi: enabling silent stream for NID %d\n",
-			per_pin->pin_nid);
+	codec_dbg(codec, "hdmi: enabling silent stream NID=0x%x\n", per_pin->pin_nid);
 
 	mutex_lock(&per_pin->lock);
 
 	if (!per_pin->channels)
-		per_pin->channels = 2;
-
-	oldval = snd_hda_codec_read(codec, per_pin->pin_nid, 0,
-			AC_VERB_GET_CONV, 0);
-	newval = (oldval & 0xF0) | 0xF;
-	snd_hda_codec_write(codec, per_pin->pin_nid, 0,
-			AC_VERB_SET_CHANNEL_STREAMID, newval);
+		per_pin->channels = INTEL_SILENTSTR_CHANNELS;
 
 	hdmi_setup_audio_infoframe(codec, per_pin, per_pin->non_pcm);
 
+	pin_idx = pin_id_to_pin_index(codec, per_pin->pin_nid, per_pin->dev_id);
+	err = hdmi_choose_cvt(codec, pin_idx, &cvt_idx);
+	if (err) {
+		codec_err(codec, "hdmi: no free converter to enable silent mode\n");
+		goto unlock_out;
+	}
+
+	per_cvt = get_cvt(spec, cvt_idx);
+	per_cvt->assigned = 1;
+	per_pin->cvt_nid = per_cvt->cvt_nid;
+	per_pin->silent_stream = true;
+
+	snd_hda_set_dev_select(codec, per_pin->pin_nid, per_pin->dev_id);
+	snd_hda_codec_write_cache(codec, per_pin->pin_nid, 0,
+				  AC_VERB_SET_CONNECT_SEL,
+				  per_pin->mux_idx);
+
+	/* configure unused pins to choose other converters */
+	pin_cvt_fixup(codec, per_pin, 0);
+
+	snd_hdac_sync_audio_rate(&codec->core, per_pin->pin_nid,
+				 per_pin->dev_id, INTEL_SILENTSTR_RATE);
+
+	/* trigger silent stream generation in hw */
+	format = snd_hdac_calc_stream_format(INTEL_SILENTSTR_RATE,
+					     per_pin->channels,
+					     INTEL_SILENTSTR_FORMAT,
+					     snd_pcm_format_width(INTEL_SILENTSTR_FORMAT),
+					     0);
+	snd_hda_codec_setup_stream(codec, per_pin->cvt_nid, INTEL_SILENTSTR_TAG, 0xf, format);
+	usleep_range(100, 200);
+	snd_hda_codec_setup_stream(codec, per_pin->cvt_nid, INTEL_SILENTSTR_TAG, 0, format);
+
+ unlock_out:
 	mutex_unlock(&per_pin->lock);
 }
 
+static void silent_stream_disable(struct hda_codec *codec,
+				  struct hdmi_spec_per_pin *per_pin)
+{
+	struct hdmi_spec_per_cvt *per_cvt = NULL;
+	struct hdmi_spec *spec = codec->spec;
+	int cvt_idx;
+
+	if (!per_pin->silent_stream)
+		return;
+
+	codec_dbg(codec, "HDMI: disable silent stream on cvt-NID=0x%x\n", per_pin->cvt_nid);
+	cvt_idx = cvt_nid_to_cvt_index(codec, per_pin->cvt_nid);
+	per_cvt = get_cvt(spec, cvt_idx);
+	WARN_ON(!per_cvt);
+	if (per_cvt)
```

Comment:
could do `if (!WARN_ON(!per_cvt))` but maybe that'd be a bit too many negations

Reply:
True. Also get_cvt() doesn't return a null, so I'll rework this in any case.